File: src/render_fb.c
// SPDX-License-Identifier: MIT
/*
 * render_fb.c - framebuffer render backend.
 *
 * Renders the overlay directly onto /dev/fb0 (or a user-supplied path).
 * Before the first draw the panel region is saved to a heap buffer; it is
 * restored when the overlay is hidden.  Pixel format (RGB565 / XRGB8888 /
 * …) is detected at runtime via FBIOGET_VSCREENINFO.
 */

#include <errno.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <linux/fb.h>
#include <sys/ioctl.h>
#include <sys/mman.h>

#include "font8x16.h"
#include "keyoverlay.h"
#include "render_fb.h"

/* ------------------------------------------------------------------ */
/* Internal framebuffer state                                         */
/* ------------------------------------------------------------------ */

struct fb {
	int      fd;
	uint8_t *mem;
	size_t   map_size;
	uint32_t xres, yres;
	uint32_t line_length;
	uint32_t bpp;           /* bytes per pixel */
	struct fb_var_screeninfo var;
};

struct fb_backend {
	render_backend  base;   /* MUST be first */
	struct fb       fb;
	struct panel    panel;
	uint8_t        *backup; /* saved pixels under the panel rectangle */
	bool            shown;
};

/* ------------------------------------------------------------------ */
/* Color packing                                                       */
/* ------------------------------------------------------------------ */

static uint32_t pack_color(const struct fb *fb, struct rgb c)
{
	uint32_t r = c.r >> (8 - fb->var.red.length);
	uint32_t g = c.g >> (8 - fb->var.green.length);
	uint32_t b = c.b >> (8 - fb->var.blue.length);
	return (r << fb->var.red.offset)   |
	       (g << fb->var.green.offset) |
	       (b << fb->var.blue.offset);
}

/* ------------------------------------------------------------------ */
/* Pixel / rectangle primitives                                        */
/* ------------------------------------------------------------------ */

static inline void put_pixel(struct fb *fb, int x, int y, uint32_t v)
{
	if (x < 0 || y < 0 || (uint32_t)x >= fb->xres || (uint32_t)y >= fb->yres)
		return;
	uint8_t *p = fb->mem + (size_t)y * fb->line_length + (size_t)x * fb->bpp;
	if (fb->bpp == 2)
		*(uint16_t *)p = (uint16_t)v;
	else
		*(uint32_t *)p = v;
}

static void fill_rect(struct fb *fb, int x, int y, int w, int h, uint32_t v)
{
	for (int yy = y; yy < y + h; yy++)
		for (int xx = x; xx < x + w; xx++)
			put_pixel(fb, xx, yy, v);
}

static void draw_rect_outline(struct fb *fb, int x, int y, int w, int h,
			      uint32_t v)
{
	for (int xx = x; xx < x + w; xx++) {
		put_pixel(fb, xx, y,         v);
		put_pixel(fb, xx, y + h - 1, v);
	}
	for (int yy = y; yy < y + h; yy++) {
		put_pixel(fb, x,         yy, v);
		put_pixel(fb, x + w - 1, yy, v);
	}
}

/* ------------------------------------------------------------------ */
/* Region save / restore                                               */
/* ------------------------------------------------------------------ */

static void region_save(struct fb *fb, const struct panel *p, uint8_t *buf)
{
	for (int yy = 0; yy < p->h; yy++) {
		const uint8_t *src = fb->mem +
			(size_t)(p->y + yy) * fb->line_length +
			(size_t)p->x * fb->bpp;
		uint8_t *dst = buf + (size_t)yy * p->w * fb->bpp;
		memcpy(dst, src, (size_t)p->w * fb->bpp);
	}
}

static void region_restore(struct fb *fb, const struct panel *p,
			    const uint8_t *buf)
{
	for (int yy = 0; yy < p->h; yy++) {
		uint8_t *dst = fb->mem +
			(size_t)(p->y + yy) * fb->line_length +
			(size_t)p->x * fb->bpp;
		const uint8_t *src = buf + (size_t)yy * p->w * fb->bpp;
		memcpy(dst, src, (size_t)p->w * fb->bpp);
	}
}

/* ------------------------------------------------------------------ */
/* Glyph / text rendering                                              */
/* ------------------------------------------------------------------ */

static void draw_glyph(struct fb *fb, int x, int y, char ch, uint32_t fg)
{
	if ((unsigned char)ch < FONT_FIRST || (unsigned char)ch > FONT_LAST)
		ch = '?';
	const unsigned char *g =
		&font8x16[((unsigned char)ch - FONT_FIRST) * FONT_H];
	for (int row = 0; row < FONT_H; row++) {
		unsigned char bits = g[row];
		for (int col = 0; col < FONT_W; col++) {
			if (!(bits & (0x80 >> col)))
				continue;
			fill_rect(fb, x + col * FONT_SCALE, y + row * FONT_SCALE,
				  FONT_SCALE, FONT_SCALE, fg);
		}
	}
}

static void draw_text(struct fb *fb, int x, int y, const char *s, uint32_t fg)
{
	for (; *s; s++, x += GLYPH_W)
		draw_glyph(fb, x, y, *s, fg);
}

/* ------------------------------------------------------------------ */
/* Layout rendering                                                    */
/* ------------------------------------------------------------------ */

static void draw_layout(struct fb_backend *b, const Layout *L)
{
	struct fb   *fb = &b->fb;
	struct panel *p  = &b->panel;

	uint32_t col_panel      = pack_color(fb, COL_PANEL);
	uint32_t col_border     = pack_color(fb, COL_BORDER);
	uint32_t col_cell       = pack_color(fb, COL_CELL);
	uint32_t col_cell_empty = pack_color(fb, COL_CELL_EMPTY);
	uint32_t col_text       = pack_color(fb, COL_TEXT);
	uint32_t col_title      = pack_color(fb, COL_TITLE);

	fill_rect(fb, p->x, p->y, p->w, p->h, col_panel);
	draw_rect_outline(fb, p->x,     p->y,     p->w,     p->h,     col_border);
	draw_rect_outline(fb, p->x + 1, p->y + 1, p->w - 2, p->h - 2, col_border);

	int content_x = p->x + PANEL_PAD;
	int y         = p->y + PANEL_PAD;

	draw_text(fb, content_x, y, L->title, col_title);
	y += GLYPH_H + TITLE_GAP;

	for (int ri = 0; ri < L->nrows; ri++) {
		const Row *r = &L->rows[ri];
		int x = content_x + (r->indent_half * MIN_CELL_W) / 2;
		for (int ci = 0; ci < r->n; ci++) {
			const char *label = r->label[ci];
			int   cw    = cell_width(label);
			bool  empty = (label[0] == '\0');
			fill_rect(fb, x, y, cw, CELL_H,
				  empty ? col_cell_empty : col_cell);
			if (!empty) {
				int tw = (int)strlen(label) * GLYPH_W;
				int tx = x + (cw - tw) / 2;
				int ty = y + (CELL_H - GLYPH_H) / 2;
				draw_text(fb, tx, ty, label, col_text);
			}
			x += cw + HGAP;
		}
		y += CELL_H + VGAP;
	}
}

/* ------------------------------------------------------------------ */
/* Backend vtable                                                      */
/* ------------------------------------------------------------------ */

static void fb_show(render_backend *base, const Layout *L)
{
	struct fb_backend *b = (struct fb_backend *)base;
	if (!b->shown)
		region_save(&b->fb, &b->panel, b->backup);
	draw_layout(b, L);
	b->shown = true;
}

static void fb_hide(render_backend *base)
{
	struct fb_backend *b = (struct fb_backend *)base;
	if (!b->shown)
		return;
	region_restore(&b->fb, &b->panel, b->backup);
	b->shown = false;
}

static void fb_close(render_backend *base)
{
	struct fb_backend *b = (struct fb_backend *)base;
	struct fb *fb = &b->fb;
	free(b->backup);
	if (fb->mem && fb->mem != MAP_FAILED)
		munmap(fb->mem, fb->map_size);
	if (fb->fd >= 0)
		close(fb->fd);
	free(b);
}

/* ------------------------------------------------------------------ */
/* Constructor                                                         */
/* ------------------------------------------------------------------ */

render_backend *render_fb_create(const char *fbpath, int verbose)
{
	struct fb_backend *b = calloc(1, sizeof(*b));
	if (!b)
		return NULL;

	b->base.show  = fb_show;
	b->base.hide  = fb_hide;
	b->base.close = fb_close;

	struct fb *fb = &b->fb;
	struct fb_fix_screeninfo fix;

	fb->fd = open(fbpath, O_RDWR);
	if (fb->fd < 0) {
		fprintf(stderr, "keyoverlay: open %s: %s\n", fbpath, strerror(errno));
		goto err_free;
	}
	if (ioctl(fb->fd, FBIOGET_VSCREENINFO, &fb->var) < 0 ||
	    ioctl(fb->fd, FBIOGET_FSCREENINFO, &fix) < 0) {
		fprintf(stderr, "keyoverlay: FBIOGET_*SCREENINFO: %s\n",
			strerror(errno));
		goto err_close;
	}

	fb->xres        = fb->var.xres;
	fb->yres        = fb->var.yres;
	fb->line_length = fix.line_length;
	fb->bpp         = fb->var.bits_per_pixel / 8;
	if (fb->bpp != 2 && fb->bpp != 4) {
		fprintf(stderr, "keyoverlay: unsupported bpp %u\n",
			fb->var.bits_per_pixel);
		goto err_close;
	}

	fb->map_size = fix.smem_len
		? fix.smem_len
		: (size_t)fb->line_length * fb->yres;
	fb->mem = mmap(NULL, fb->map_size, PROT_READ | PROT_WRITE, MAP_SHARED,
		       fb->fd, 0);
	if (fb->mem == MAP_FAILED) {
		fprintf(stderr, "keyoverlay: mmap: %s\n", strerror(errno));
		goto err_close;
	}

	if (verbose)
		fprintf(stderr, "keyoverlay: fb %ux%u %ubpp\n",
			fb->xres, fb->yres, fb->var.bits_per_pixel);

	b->panel = compute_panel((int)fb->xres, (int)fb->yres);

	size_t region_sz = (size_t)b->panel.w * b->panel.h * fb->bpp;
	b->backup = malloc(region_sz);
	if (!b->backup) {
		fprintf(stderr, "keyoverlay: out of memory for region backup\n");
		goto err_unmap;
	}

	return &b->base;

err_unmap:
	munmap(fb->mem, fb->map_size);
err_close:
	close(fb->fd);
err_free:
	free(b);
	return NULL;
}

File: src/render_fb.zig
// SPDX-License-Identifier: MIT
//
// render_fb.zig - framebuffer render backend.
//
// Renders the overlay directly onto /dev/fb0 (or a user-supplied path).
// Before the first draw the panel region is saved to a heap buffer; it is
// restored when the overlay is hidden.  Pixel format (RGB565 / XRGB8888 / …)
// is detected at runtime via FBIOGET_VSCREENINFO.

const std = @import("std");
const lx = @import("linux.zig");
const c = lx.c;
const font = @import("font8x16.zig");
const lo = @import("layout.zig");

const Layout = lo.Layout;
const Panel = lo.Panel;
const Rgb = lo.Rgb;
const Backend = @import("render.zig").Backend;

fn errno() c_int {
    return std.c._errno().*;
}

pub const FbBackend = struct {
    alloc: std.mem.Allocator,
    fd: c_int,
    mem: []u8,
    map_size: usize,
    xres: u32,
    yres: u32,
    line_length: u32,
    bpp: u32, // bytes per pixel
    var_info: c.struct_fb_var_screeninfo,
    panel: Panel,
    backup: []u8,
    shown: bool,

    const vtable = Backend.VTable{
        .show = showThunk,
        .hide = hideThunk,
        .deinit = deinitThunk,
    };

    pub fn backend(self: *FbBackend) Backend {
        return .{ .ptr = self, .vtable = &vtable };
    }

    // -------------------------------------------------------------- //
    // Color packing                                                   //
    // -------------------------------------------------------------- //

    fn packColor(self: *const FbBackend, col: Rgb) u32 {
        const v = &self.var_info;
        const r = @as(u32, col.r) >> @intCast(8 - v.red.length);
        const g = @as(u32, col.g) >> @intCast(8 - v.green.length);
        const b = @as(u32, col.b) >> @intCast(8 - v.blue.length);
        return (r << @intCast(v.red.offset)) |
            (g << @intCast(v.green.offset)) |
            (b << @intCast(v.blue.offset));
    }

    // -------------------------------------------------------------- //
    // Pixel / rectangle primitives                                    //
    // -------------------------------------------------------------- //

    fn putPixel(self: *FbBackend, x: i32, y: i32, v: u32) void {
        if (x < 0 or y < 0 or
            x >= @as(i32, @intCast(self.xres)) or
            y >= @as(i32, @intCast(self.yres)))
            return;
        const off = @as(usize, @intCast(y)) * self.line_length +
            @as(usize, @intCast(x)) * self.bpp;
        const p = self.mem.ptr + off;
        if (self.bpp == 2) {
            @as(*align(1) u16, @ptrCast(p)).* = @truncate(v);
        } else {
            @as(*align(1) u32, @ptrCast(p)).* = v;
        }
    }

    fn fillRect(self: *FbBackend, x: i32, y: i32, w: i32, h: i32, v: u32) void {
        var yy = y;
        while (yy < y + h) : (yy += 1) {
            var xx = x;
            while (xx < x + w) : (xx += 1)
                self.putPixel(xx, yy, v);
        }
    }

    fn drawRectOutline(self: *FbBackend, x: i32, y: i32, w: i32, h: i32, v: u32) void {
        var xx = x;
        while (xx < x + w) : (xx += 1) {
            self.putPixel(xx, y, v);
            self.putPixel(xx, y + h - 1, v);
        }
        var yy = y;
        while (yy < y + h) : (yy += 1) {
            self.putPixel(x, yy, v);
            self.putPixel(x + w - 1, yy, v);
        }
    }

    // -------------------------------------------------------------- //
    // Region save / restore                                           //
    // -------------------------------------------------------------- //

    fn regionSave(self: *FbBackend) void {
        const p = self.panel;
        const pw: usize = @intCast(p.w);
        const ph: usize = @intCast(p.h);
        const px: usize = @intCast(p.x);
        const py: usize = @intCast(p.y);
        const row_bytes = pw * self.bpp;
        var yy: usize = 0;
        while (yy < ph) : (yy += 1) {
            const src = (py + yy) * self.line_length + px * self.bpp;
            const dst = yy * row_bytes;
            @memcpy(self.backup[dst .. dst + row_bytes], self.mem[src .. src + row_bytes]);
        }
    }

    fn regionRestore(self: *FbBackend) void {
        const p = self.panel;
        const pw: usize = @intCast(p.w);
        const ph: usize = @intCast(p.h);
        const px: usize = @intCast(p.x);
        const py: usize = @intCast(p.y);
        const row_bytes = pw * self.bpp;
        var yy: usize = 0;
        while (yy < ph) : (yy += 1) {
            const dst = (py + yy) * self.line_length + px * self.bpp;
            const src = yy * row_bytes;
            @memcpy(self.mem[dst .. dst + row_bytes], self.backup[src .. src + row_bytes]);
        }
    }

    // -------------------------------------------------------------- //
    // Glyph / text rendering                                          //
    // -------------------------------------------------------------- //

    fn drawGlyph(self: *FbBackend, x: i32, y: i32, ch: u8, fg: u32) void {
        var cc = ch;
        if (cc < font.FONT_FIRST or cc > font.FONT_LAST) cc = '?';
        const base = (@as(usize, cc) - font.FONT_FIRST) * font.FONT_H;
        var row: usize = 0;
        while (row < font.FONT_H) : (row += 1) {
            const bits = font.font8x16[base + row];
            var col: usize = 0;
            while (col < font.FONT_W) : (col += 1) {
                if (bits & (@as(u8, 0x80) >> @intCast(col)) == 0) continue;
                self.fillRect(
                    x + @as(i32, @intCast(col)) * lo.FONT_SCALE,
                    y + @as(i32, @intCast(row)) * lo.FONT_SCALE,
                    lo.FONT_SCALE,
                    lo.FONT_SCALE,
                    fg,
                );
            }
        }
    }

    fn drawText(self: *FbBackend, x: i32, y: i32, s: []const u8, fg: u32) void {
        var cx = x;
        for (s) |ch| {
            self.drawGlyph(cx, y, ch, fg);
            cx += lo.GLYPH_W;
        }
    }

    // -------------------------------------------------------------- //
    // Layout rendering                                                //
    // -------------------------------------------------------------- //

    fn drawLayout(self: *FbBackend, l: *const Layout) void {
        const p = self.panel;

        const col_panel = self.packColor(lo.COL_PANEL);
        const col_border = self.packColor(lo.COL_BORDER);
        const col_cell = self.packColor(lo.COL_CELL);
        const col_cell_empty = self.packColor(lo.COL_CELL_EMPTY);
        const col_text = self.packColor(lo.COL_TEXT);
        const col_title = self.packColor(lo.COL_TITLE);

        self.fillRect(p.x, p.y, p.w, p.h, col_panel);
        self.drawRectOutline(p.x, p.y, p.w, p.h, col_border);
        self.drawRectOutline(p.x + 1, p.y + 1, p.w - 2, p.h - 2, col_border);

        const content_x = p.x + lo.PANEL_PAD;
        var y = p.y + lo.PANEL_PAD;

        self.drawText(content_x, y, l.title, col_title);
        y += lo.GLYPH_H + lo.TITLE_GAP;

        for (l.rows) |*r| {
            var x = content_x + @divTrunc(r.indent_half * lo.MIN_CELL_W, 2);
            for (r.label) |label| {
                const cw = lo.cellWidth(label);
                const empty = label.len == 0;
                self.fillRect(x, y, cw, lo.CELL_H, if (empty) col_cell_empty else col_cell);
                if (!empty) {
                    const tw = @as(i32, @intCast(label.len)) * lo.GLYPH_W;
                    const tx = x + @divTrunc(cw - tw, 2);
                    const ty = y + @divTrunc(lo.CELL_H - lo.GLYPH_H, 2);
                    self.drawText(tx, ty, label, col_text);
                }
                x += cw + lo.HGAP;
            }
            y += lo.CELL_H + lo.VGAP;
        }
    }

    // -------------------------------------------------------------- //
    // Backend vtable                                                  //
    // -------------------------------------------------------------- //

    fn show(self: *FbBackend, l: *const Layout) void {
        if (!self.shown) self.regionSave();
        self.drawLayout(l);
        self.shown = true;
    }

    fn hide(self: *FbBackend) void {
        if (!self.shown) return;
        self.regionRestore();
        self.shown = false;
    }

    fn deinit(self: *FbBackend) void {
        const alloc = self.alloc;
        alloc.free(self.backup);
        _ = lx.munmap(@ptrCast(self.mem.ptr), self.map_size);
        _ = lx.close(self.fd);
        alloc.destroy(self);
    }

    fn showThunk(ptr: *anyopaque, l: *const Layout) void {
        show(@ptrCast(@alignCast(ptr)), l);
    }
    fn hideThunk(ptr: *anyopaque) void {
        hide(@ptrCast(@alignCast(ptr)));
    }
    fn deinitThunk(ptr: *anyopaque) void {
        deinit(@ptrCast(@alignCast(ptr)));
    }
};

// ------------------------------------------------------------------ //
// Constructor                                                         //
// ------------------------------------------------------------------ //

pub fn create(alloc: std.mem.Allocator, fbpath: [*:0]const u8, verbose: bool) ?Backend {
    const self = alloc.create(FbBackend) catch return null;
    errdefer alloc.destroy(self);

    const fd = lx.open(fbpath, lx.O_RDWR);
    if (fd < 0) {
        std.debug.print("keyoverlay: open {s}: errno {d}\n", .{ fbpath, errno() });
        return null;
    }
    errdefer _ = lx.close(fd);

    var var_info: c.struct_fb_var_screeninfo = undefined;
    var fix: c.struct_fb_fix_screeninfo = undefined;
    if (lx.ioctl(fd, c.FBIOGET_VSCREENINFO, &var_info) < 0 or
        lx.ioctl(fd, c.FBIOGET_FSCREENINFO, &fix) < 0)
    {
        std.debug.print("keyoverlay: FBIOGET_*SCREENINFO failed\n", .{});
        return null;
    }

    const xres = var_info.xres;
    const yres = var_info.yres;
    const line_length = fix.line_length;
    const bpp = var_info.bits_per_pixel / 8;
    if (bpp != 2 and bpp != 4) {
        std.debug.print("keyoverlay: unsupported bpp {d}\n", .{var_info.bits_per_pixel});
        return null;
    }

    const map_size: usize = if (fix.smem_len != 0)
        fix.smem_len
    else
        @as(usize, line_length) * yres;

    const raw = lx.mmap(null, map_size, lx.PROT_READ | lx.PROT_WRITE, lx.MAP_SHARED, fd, 0);
    if (raw == lx.MAP_FAILED) {
        std.debug.print("keyoverlay: mmap failed\n", .{});
        return null;
    }
    const mem = @as([*]u8, @ptrCast(raw))[0..map_size];
    errdefer _ = lx.munmap(@ptrCast(mem.ptr), map_size);

    if (verbose)
        std.debug.print("keyoverlay: fb {d}x{d} {d}bpp\n", .{ xres, yres, var_info.bits_per_pixel });

    const panel = lo.computePanel(@intCast(xres), @intCast(yres));

    const region_sz = @as(usize, @intCast(panel.w)) * @as(usize, @intCast(panel.h)) * bpp;
    const backup = alloc.alloc(u8, region_sz) catch {
        std.debug.print("keyoverlay: out of memory for region backup\n", .{});
        return null;
    };

    self.* = .{
        .alloc = alloc,
        .fd = fd,
        .mem = mem,
        .map_size = map_size,
        .xres = xres,
        .yres = yres,
        .line_length = line_length,
        .bpp = bpp,
        .var_info = var_info,
        .panel = panel,
        .backup = backup,
        .shown = false,
    };
    return self.backend();
}
